src/serac/physics/utilities/functional/functional.hpp

```diff
@@ -135,65 +154,47 @@ class Functional<test(trial)> : public mfem::Operator {
+    auto num_elements = domain.GetNE();
+    SLIC_ERROR_IF(num_elements == 0, "Mesh has no elements");
 
-      auto dim = domain.Dimension();
-      for (int e = 0; e < num_elements; e++) {
-        SLIC_ERROR_IF(domain.GetElementType(e) != supported_types[dim], "Mesh contains unsupported element type");
-      }
+    SLIC_ERROR_IF(dim != domain.Dimension(), "Error: invalid mesh dimension for domain integral");
+    for (int e = 0; e < num_elements; e++) {
+      SLIC_ERROR_IF(domain.GetElementType(e) != supported_types[dim], "Mesh contains unsupported element type");
+    }
 
-      const mfem::FiniteElement&   el = *test_space_->GetFE(0);
-      const mfem::IntegrationRule& ir = mfem::IntRules.Get(el.GetGeomType(), el.GetOrder() * 2);
+    const mfem::FiniteElement&   el = *test_space_->GetFE(0);
+    const mfem::IntegrationRule& ir = mfem::IntRules.Get(el.GetGeomType(), el.GetOrder() * 2);
 
-      constexpr auto flags = mfem::GeometricFactors::COORDINATES | mfem::GeometricFactors::JACOBIANS;
-      auto           geom  = domain.GetGeometricFactors(ir, flags);
-      domain_integrals_.emplace_back(num_elements, geom->J, geom->X, Dimension<geometry_dim>{},
-                                     Dimension<spatial_dim>{}, integrand);
-      return;
-    }
-#ifdef ENABLE_BOUNDARY_INTEGRALS
-    else if constexpr ((geometry_dim + 1) == spatial_dim) {
-      // TODO: fix mfem::FaceGeometricFactors
-      constexpr bool always_false = (geometry_dim == spatial_dim);
-      static_assert(always_false,
-                    "unsupported integral dimensionality due to unimplemented features in mfem::FaceGeometricFactors");
-
-      auto num_boundary_elements = domain.GetNBE();
-      SLIC_ERROR_IF(num_boundary_elements == 0, "Mesh has no boundary elements");
-
-      auto dim = domain.Dimension();
-      SLIC_ERROR_IF(dim != spatial_dim, "Error: invalid mesh dimension for integral");
-      for (int e = 0; e < num_boundary_elements; e++) {
-        SLIC_ERROR_IF(domain.GetBdrElementType(e) != supported_types[dim - 1],
-                      "Mesh contains unsupported element type");
-      }
+    constexpr auto flags = mfem::GeometricFactors::COORDINATES | mfem::GeometricFactors::JACOBIANS;
+    auto           geom  = domain.GetGeometricFactors(ir, flags);
+    domain_integrals_.emplace_back(num_elements, geom->J, geom->X, Dimension<dim>{}, integrand);
+  }
 
-      const mfem::FiniteElement&   el = *test_space_->GetBE(0);
-      const mfem::IntegrationRule& ir = mfem::IntRules.Get(el.GetGeomType(), el.GetOrder() * 2);
-      constexpr auto flags            = mfem::FaceGeometricFactors::COORDINATES | mfem::FaceGeometricFactors::JACOBIANS;
+  template <int dim, typename lambda>
+  void AddBoundaryIntegral(Dimension<dim>, lambda&& integrand, mfem::Mesh& domain)
+  {
+    // TODO: fix mfem::FaceGeometricFactors
+    auto num_boundary_elements = domain.GetNBE();
+    SLIC_ERROR_IF(num_boundary_elements == 0, "Mesh has no boundary elements");
```

Comment:
what does that mean?

Reply:
It should be `SLIC_ERROR_ROOT_IF` instead of `SLIC_ERROR_IF`. The `ROOT` means it will only display on processor 0. This can really help clean up logs for large scale runs, It's described in the docs [here](https://serac.readthedocs.io/en/latest/sphinx/dev_guide/logging.html).

Reply:
```suggestion
    SLIC_ERROR_ROOT_IF(num_boundary_elements == 0, "Mesh has no boundary elements");
```